cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java

```diff
@@ -135,8 +132,11 @@ public void run(Context context) throws IOException, InterruptedException {
       }
 
     } finally {
-      basicMapReduceContext.close();
-      basicMapReduceContext.getMetricsCollectionService().stop();
+      try {
+        basicMapReduceContext.close();
+      } finally {
+        basicMapReduceContext.getMetricsCollectionService().stop();
```

Comment:
It cannot. There are multiple places that call BasicMapReduceContext.close(), and in some context (e.g. MapReduceRuntimeService in local mode), it shares the same instance as the system metrics collection service. The current code base is pretty messy around creation and closing of the MR context object. It's part of the refactoring tracked by https://issues.cask.co/browse/CDAP-2937